5.next - Add DI support for Components

markstory opened the pull request:
This implements the somewhat frequently requested feature of having DI support for components. I've included tests at a few different layers as some critical path methods have changed.

This *should* be backwards compatible with custom constructors. Right now I'm only matching parameters based on name, I did this because it was fast and simple to write. We probably want type based checking instead and I'd like to hear folks thoughts on how we should approach this.

Refs #17231

markstory committed:
Prototype of container injection into component registry

This would unblock getting components from the DI container. This gives
existing applications a great way to onboard to the DIC and
incrementally adopt it in more places.

markstory committed:
Fix bugs and add tests

markstory commented on src/Controller/Controller.php:
```diff
@@ -223,6 +225,9 @@ public function __construct(
         if ($eventManager !== null) {
             $this->setEventManager($eventManager);
         }
+        if ($container !== null) {
+            $this->_components = new ComponentRegistry($this, $container);
+        }
```
Having `container` passed into controller constructors could go awry. It will be the first time that we have made it easy for application code to get access to the container in a direct simple way. I don't think there is a backwards compatible way to build this right now if we don't want to pass the container in. I'm ok with moving this to 6.x and cleaning it up there too.

ADmad replied:
> Having container passed into controller constructors could go awry.

Yup. I too had looked into adding DI support for components a while back and had come up with [this](https://github.com/cakephp/cakephp/compare/5.next...ADmad:cakephp:5.x-component-di). 

The difference is instead of passing the container instance to the controller I am passing in a component registry instance. It did create a bit of a chicken and egg problem between the controller and component registry which is why I had to add a method to set the controller instance to the registry later. So we can weigh the pros and cons of both approaches.

I didn't open a PR for my changes is because I question the value/need for components all together, with the availability of  traits and the ability to add middlewares for controllers. Only benefit I see is the convention based auto run event callbacks.

markstory replied:
> The difference is instead of passing the container instance to the controller I am passing in a component registry instance. It did create a bit of a chicken and egg problem between the controller and component registry which is why I had to add a method to set the controller instance to the registry later. So we can weigh the pros and cons of both approaches.

Passing ComponentRegistry to `Controller` might be the better solution, as it doesn't require changing controller's constructor.

> I didn't open a PR for my changes is because I question the value/need for components all together, with the availability of traits and the ability to add middlewares for controllers. 

Having components able to use services does give folks a way to incrementally refactor their application away from components and use services instead.

ADmad replied:
> ...as it doesn't require changing controller's constructor.

My branch does the controller's constructor and add component registry as an argument.

> Having components able to use services does give folks a way to incrementally refactor their application away from components and use services instead.

Okay, I am not against this change then.

markstory committed:
Get tests passing

Shuffle where container assignments are done so that constructing an
object doesn't have side-effects and to resolve psalm errors.

markstory committed:
Fix phpstan errors

markstory committed:
Invert the control flow better.

Take inspiration from @ADmad's approach in

https://github.com/cakephp/cakephp/compare/5.next...ADmad:cakephp:5.x-component-di

I've had to retain more complex logic to accomodate us adding a new
constructor parameter, and having already given users the go ahead in
using constructors for DI, and we can't add a new required parameter.
However, if users do update their controller constructors, it all
magically works.

I'd like to deprecate not having components in controller constructors,
in 5.2 and then we can fully remove it in the future.

markstory committed:
Fix phpcs and psalm

othercorey commented on src/Controller/ComponentRegistry.php:
```diff
@@ -63,6 +88,10 @@ public function __construct(Controller $controller)
     public function getController(): Controller
     {
+        if ($this->_Controller === null) {
+            throw new RuntimeException('Cannot getController() it has not been set yet.');
```
```suggestion
            throw new RuntimeException('Controller must be set first.');
```

ADmad commented on src/Controller/ComponentRegistry.php:
```diff
@@ -41,19 +43,42 @@ class ComponentRegistry extends ObjectRegistry implements EventDispatcherInterfa
+     * @var \Cake\Controller\Controller|null
      */
-    protected Controller $_Controller;
+    protected ?Controller $_Controller;
```
```suggestion
    protected ?Controller $_Controller = null;
```

ADmad commented on src/Controller/ComponentRegistry.php:
```diff
@@ -41,19 +43,42 @@ class ComponentRegistry extends ObjectRegistry implements EventDispatcherInterfa
+     */
+    public function __construct(?Controller $controller = null, ?ContainerInterface $container = null)
+    {
+        $this->_Controller = null;
```
```suggestion
```

ADmad commented on src/Controller/ComponentRegistry.php:
```diff
@@ -41,19 +43,42 @@ class ComponentRegistry extends ObjectRegistry implements EventDispatcherInterfa
+    protected ?Controller $_Controller;
+
+    /**
+     * @var \Cake\Core\ContainerInterface
```
```suggestion
     * @var \Cake\Core\ContainerInterface|null
```

ADmad commented on src/Controller/ComponentRegistry.php:
```diff
@@ -41,19 +43,42 @@ class ComponentRegistry extends ObjectRegistry implements EventDispatcherInterfa
+    /**
+     * @var \Cake\Core\ContainerInterface
+     */
+    protected ?ContainerInterface $container;
```
```suggestion
    protected ?ContainerInterface $container = null;
```

ADmad commented on src/Controller/ComponentRegistry.php:
```diff
@@ -115,8 +144,14 @@ protected function _create(object|string $class, string $alias, array $config):
         if (is_object($class)) {
             return $class;
         }
+        if ($this->container && $this->container->has($class)) {
```
```suggestion
        if ($this->container?->has($class)) {
```

Modern syntax :)

ADmad commented on src/Controller/ControllerFactory.php:
```diff
@@ -78,13 +78,37 @@ public function create(ServerRequestInterface $request): Controller
         }
+        $this->container->addShared(
+            ComponentRegistry::class,
+            new ComponentRegistry(null, $this->container)
```
```suggestion
            new ComponentRegistry(container: $this->container)
```

ADmad commented on src/Controller/ControllerFactory.php:
```diff
@@ -78,13 +78,37 @@ public function create(ServerRequestInterface $request): Controller
+            $hasComponents = false;
+            foreach ($constructor->getParameters() as $parameter) {
+                $paramType = $parameter->getType();
+                // TODO: In a future minor release it would be good to start requiring the components parameter
```
People won't know they need to upgrade the constructor signature if we don't show a warning.

markstory replied:
Yes, do we want to add a deprecation for this now, or in a future release? I know I usually regret letting deprecation TODOs accumulate, but also don't want to bombard developers on the first point release.

ADmad replied:
I guess we can punt the message to 5.2.

markstory replied:
That was my thinking too.

markstory committed:
Apply suggestions from code review

Co-authored-by: ADmad <ADmad@users.noreply.github.com>
Co-authored-by: othercorey <corey.taylor.fl@gmail.com>
